File: frontend/src/data/providers/registry.ts
/**
 * Provider registry — the single place that knows which providers exist.
 *
 * Adding another AMFI-derived source is a two-line change here plus a new
 * module in this folder; nothing else in the app needs to be touched.
 *
 * VITE_MFAPI_BASE_URL overrides the endpoint. That is the escape hatch if
 * MFapi.in ever stops allowing direct browser calls: point it at a pass-through
 * proxy you control that forwards the same /mf/... paths and adds permissive
 * CORS headers. It must mirror MFapi.in's URL shape and response format.
 */

import { MFApiProvider } from './mfapi'
import type { FundDataProvider } from './types'

const factories: Record<string, () => FundDataProvider> = {
  mfapi: () => new MFApiProvider(import.meta.env.VITE_MFAPI_BASE_URL),
}

const instances: Record<string, FundDataProvider> = {}

export function registerProvider(key: string, factory: () => FundDataProvider): void {
  factories[key] = factory
}

export function availableProviders(): string[] {
  return Object.keys(factories).sort()
}

export function getProvider(key?: string): FundDataProvider {
  const name = (key ?? import.meta.env.VITE_MF_PROVIDER ?? 'mfapi').toLowerCase()
  const factory = factories[name]
  if (!factory) {
    throw new Error(
      `Unknown data provider "${name}". Available: ${availableProviders().join(', ')}`,
    )
  }
  instances[name] ??= factory()
  return instances[name]
}

export function setProvider(key: string, provider: FundDataProvider): void {
  instances[key] = provider
}

File: backend/app/providers/registry.py
"""Provider registry — the single place that knows which providers exist.

Adding a second AMFI-derived source is a two-line change here plus a new module
in this package; nothing else in the application needs to be touched.
"""

from __future__ import annotations

from typing import Callable

from ..config import settings
from .base import FundDataProvider
from .mfapi import MFApiProvider

_FACTORIES: dict[str, Callable[[], FundDataProvider]] = {
    "mfapi": MFApiProvider,
}

_instances: dict[str, FundDataProvider] = {}


def register_provider(key: str, factory: Callable[[], FundDataProvider]) -> None:
    _FACTORIES[key] = factory


def available_providers() -> list[str]:
    return sorted(_FACTORIES)


def get_provider(key: str | None = None) -> FundDataProvider:
    key = (key or settings.provider).lower()
    if key not in _FACTORIES:
        raise KeyError(
            f"Unknown data provider {key!r}. Available: {', '.join(available_providers())}"
        )
    if key not in _instances:
        _instances[key] = _FACTORIES[key]()
    return _instances[key]


def set_provider(key: str, provider: FundDataProvider) -> None:
    """Override the live instance (used by tests and by future hot-swapping)."""
    _instances[key] = provider


async def close_providers() -> None:
    for provider in list(_instances.values()):
        await provider.aclose()
    _instances.clear()
